src/librustdoc/html/render.rs

```diff
@@ -2304,6 +2363,53 @@ fn make_item_keywords(it: &clean::Item) -> String {
     format!("{}, {}", get_basic_keywords(), it.name.as_ref().unwrap())
 }
 
+fn get_index_search_type(item: &clean::Item,
+                         parent: Option<String>) -> Option<IndexItemFunctionType> {
+    match item.inner {
+        clean::FunctionItem(ref f) => Some(get_index_fn_type(&f.decl, None)),
+        clean::MethodItem(ref m) => Some(get_index_fn_type(&m.decl, parent)),
+        clean::TyMethodItem(ref m) => Some(get_index_fn_type(&m.decl, parent)),
+        _ => None
+    }
+}
+
+fn get_index_fn_type(decl: &clean::FnDecl, parent: Option<String>) -> IndexItemFunctionType {
+    let mut inputs = Vec::new();
+    // Consider `self` an argument as well.
+    match parent {
+        Some(name) => inputs.push(Type { name: name.into_ascii_lowercase() }),
+        None => {}
+    };
+    for arg in decl.inputs.values.iter() {
+        inputs.push(get_index_type(&arg.type_));
+    }
+
+    let output = match decl.output {
+        clean::FunctionRetTy::Return(ref return_type) => Some(get_index_type(return_type)),
+        _ => None
+    };
+
+    IndexItemFunctionType { inputs: inputs, output: output }
+}
+
+fn get_index_type(clean_type: &clean::Type) -> Type {
+    Type { name: get_index_type_name(clean_type).into_ascii_lowercase() }
+}
+
+fn get_index_type_name(clean_type: &clean::Type) -> String {
+    match clean_type {
+        &clean::ResolvedPath { ref path, .. } => {
+            let segments = &path.segments;
+            segments[segments.len() - 1].name.clone()
+        },
+        &clean::Generic(ref s) => s.clone(),
+        &clean::Primitive(ref p) => format!("{:?}", p),
+        &clean::BorrowedRef { ref type_, .. } => get_index_type_name(&*(type_.clone())),
```

Comment:
I think this can just be `get_index_type_name(type_)` (no need to clone or add `&` or `*`